### PR TITLE
Adds a VV Menu for Manipulating Reagent Holders

### PR DESCRIPTION
**Pull request by Arthri**
## What Does This PR Do
- Added a an "Edit Reagent" button to all atom VVs
- Added a `get_reagent_by_id` proc to `reagents_holder.dm` to get a reagent inside the container given a reagent ID

## Why It's Good For The Game
Makes adding/removing reagents from a container easier for admins

## Images of changes
https://github.com/ParadiseSS13/Paradise/assets/41360489/d9c4b40f-c61b-4308-b43e-9d45e4215d4f

## Testing
Pressed buttons and stuff to make sure that the UI works

## Changelog
:cl: cdui
add: [ADMIN] /tg/ui menu to atom VVs for manipulating reagents
/:cl:

### CODE REVIEW

File: code/datums/datumvars.dm [hunk 1]
```diff
@@ -948,6 +948,13 @@
 					log_admin("[key_name(usr)] has added [amount] units of [chosen_id] to \the [A]")
 					message_admins("<span class='notice'>[key_name(usr)] has added [amount] units of [chosen_id] to \the [A]</span>")
 
+	else if(href_list["editreagents"])
+		if(!check_rights(R_DEBUG|R_ADMIN))	return
```
Comment (DGamerL) [inline code/datums/datumvars.dm:952]:
```suggestion
		if(!check_rights(R_DEBUG|R_ADMIN))	
			return
```

File: code/modules/admin/reagents_editor.dm [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+/datum/reagents_editor/ui_interact(mob/user, datum/tgui/ui)
+	ui = SStgui.try_update_ui(user, src, ui)
+	if(!ui)
+		ui = new(user, src, "ReagentsEditor")
+		ui.open()
+		ui.set_autoupdate(FALSE)
+
+/datum/reagents_editor/ui_close(mob/user)
+	var/open_uis = SStgui.open_uis_by_src[src.UID()]
+	if (isnull(open_uis) || !islist(open_uis) || length(open_uis) <= 1)
```
Comment (DGamerL) [inline code/modules/admin/reagents_editor.dm:21]:
```suggestion
	if(isnull(open_uis) || !islist(open_uis) || length(open_uis) <= 1)
```

File: code/modules/admin/reagents_editor.dm [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+// This interface intentionally emulates VV.
+// It should, therefore, doesn't place restrictions on any actions, which includes but
+// is not limited to: adding a reagent which overfills the container and adding blood
+// with a non-existent blood type. It may also do things unconventionally, such as
+// directly appending reagents to the list rather than using reagents.add_reagent to
+// bypass reagent reactions.
+/datum/reagents_editor/ui_act(action, list/params, datum/tgui/ui, datum/ui_state/state)
+	. = ..()
+
+	if (.)
```
Comment (DGamerL) [inline code/modules/admin/reagents_editor.dm:56]:
```suggestion
	if(.)
```

File: code/modules/client/client_procs.dm [hunk 4]
```diff
@@ -1243,6 +1243,17 @@
 			debug_variables(stat_item)
 			message_admins("Admin [key_name_admin(usr)] is debugging the [stat_item] [class].")
 
+/client/proc/try_open_reagent_editor(atom/target)
+	var/target_UID = target.UID()
+	var/datum/reagents_editor/editor
+	// editors is static, it can be accessed using a null reference
+	editor = editor.editors[target_UID]
+	if (!editor)
```
Comment (DGamerL) [inline code/modules/client/client_procs.dm:1251]:
```suggestion
	if(!editor)
```

File: code/modules/reagents/chemistry/reagents_holder.dm [hunk 5]
```diff
@@ -810,6 +810,15 @@
...
 	. = locate(type) in reagent_list
 
+/datum/reagents/proc/get_reagent_by_id(id)
+	var/list/cached_reagents = reagent_list
+	for(var/A in cached_reagents)
+		var/datum/reagent/R = A
+		if(R.id == id)
+			return R
+
+	return null
```
Comment (DGamerL) [inline code/modules/reagents/chemistry/reagents_holder.dm:820]:
```suggestion
	return 
```

**Comment by meow20** [comment]
Thats absolutely insane wow :O

**Review by S34NW — APPROVED** [review]

**Comment by Burzah** [comment]
!merge_upstream

**Comment by lewcc** [comment]
!merge_upstream

**Comment by S34NW** [comment]
!merge_upstream

**Pull request merged**